transactions.go

```diff
@@ -12,29 +15,50 @@ func NewTransactions() *Transactions {
 }
 
 func (t *Transactions) Add(tx *Transaction) {
+	t.mu.Lock()
 	t.buffer[tx.ID] = tx
+	t.mu.Unlock()
 }
 
 // Will return nil if the transaction does not exist.
 func (t *Transactions) Get(id TransactionID) *Transaction {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+
 	return t.buffer[id]
 }
 
 func (t *Transactions) Has(id TransactionID) bool {
-	_, exist := t.buffer[id]
-	return exist
+	return t.Get(id) != nil
```

Comment:
Curious, is it cheaper to do `_, exist := t.buffer[id]; return exist` over `return t.buffer[id] != nil`?

Reply:
They both access the map once and they'd do the same thing, I don't think so.